rpc/client.go

```diff
@@ -306,6 +324,9 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 			if err = c.heartbeat(retryOpts.Closer); err != nil {
 				setUnhealthy()
 				log.Warning(err)
+				if c.maybeClose(retryOpts.Closer) {
```

Comment:
I'm not sure if this was necessary. In one test failure the client noticed it was closed (`err == errClosed`) but then hit the retry loop again before actually closing. That seemed wrong.